File: Boj/11724.py
N, M = map(int, input().split())

graph = [[] for _ in range(0, N)]
visited = [False for _ in range(N)]
cnt = 0

for i in range(0, M):
    u, v = map(lambda x: int(x)-1, input().split())
    graph[u].append(v)
    graph[v].append(u)


def dfs(graph, start, visited):
    visited[start] = True

    for i in graph[start]:
        if not visited[i]:
            dfs(graph, i, visited)


for i in range(0, len(visited)):
    if visited[i] == False:
        cnt += 1
        dfs(graph, i, visited)

print(cnt)
# 변수명 추천
# 인접 행렬 노드개수 많을 수록 메모리 낭비
# 인접 리스트 메모리 효율 노드연결 정보 얻는 속도 느리다

File: Boj_kt/11724.kt
fun main() {
    val (v, edge) = readln().split(" ").map { it.toInt() }
    val graph = Array<ArrayList<Int>>(v + 1) { arrayListOf() }
    val visited = BooleanArray(v + 1) { false }
    var cnt = 0

    repeat(edge) {
        val (v1, v2) = readln().split(" ").map { it.toInt() }
        graph[v1].add(v2)
        graph[v2].add(v1)
    }

    for (i in 1..v) {
        if (!visited[i]) {
            cnt++
            dfs(graph, i, visited)
        }
    }
    print(cnt)
}

fun dfs(graph: Array<ArrayList<Int>>, v: Int, visited: BooleanArray) {
    visited[v] = true
    graph[v].forEach {
        if (!visited[it]) {
            dfs(graph, it, visited)
        }
    }
}